include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -33,97 +33,98 @@ struct counting
                 template __call <SharedAccessor>(value, sdata);
     }
 
-    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    uint32_t toroidal_histogram_add(uint32_t tid, uint32_t sum, NBL_REF_ARG(HistogramAccessor) histogram, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
+    {
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        sdata.set(tid % GroupSize, sum);
+        uint32_t shifted_tid = (tid + glsl::gl_SubgroupSize() * params.workGroupIndex) % GroupSize;
+
+        sdata.workgroupExecutionAndMemoryBarrier();
+
+        return histogram.atomicAdd((tid / GroupSize) * GroupSize + shifted_tid, sdata.get(shifted_tid));
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> params)
     {
         uint32_t tid = workgroup::SubgroupContiguousIndex();
-        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
-        uint32_t baseIndex = (glsl::gl_WorkGroupID().x * GroupSize) * data.elementsPerWT;
 
-        for (int i = 0; i < buckets_per_thread; i++) {
-            uint32_t prev_bucket_count = GroupSize * i;
-            sdata.set(prev_bucket_count + tid, 0);
+        for (; tid < KeyBucketCount; tid += GroupSize) {
+            sdata.set(tid, 0);
         }
 
         sdata.workgroupExecutionAndMemoryBarrier();
 
-        for (int i = 0; i < data.elementsPerWT; i++)
+        uint32_t index = params.workGroupIndex * GroupSize * params.elementsPerWT + tid % GroupSize;
```

Comment:
why areyou using % operator? This is literally the most expensive operator on the GPU (unless when the rhs is a Power of Two)